clang/test/CXX/drs/cwg26xx.cpp

```diff
@@ -24,6 +44,103 @@ using enum E;
 #endif
 }
 
+namespace cwg2627 { // cwg2627: 19
+#if __cplusplus >= 202002L
+struct C {
+  long long i : 8;
+  friend auto operator<=>(C, C) = default;
+};
+
+void f() {
+  C x{1}, y{2};
+  static_cast<void>(x <=> y);
+  static_cast<void>(x.i <=> y.i);
+}
+
+template<typename T>
+struct CDependent {
+  T i : 8;
+  friend auto operator<=>(CDependent, CDependent) = default;
+};
+
+template<typename T>
+concept three_way_comparable = requires(T t) { { t <=> t }; };
+template<typename T>
+concept bf_three_way_comparable = requires(T t) { { t.i <=> t.i }; };
+static_assert(three_way_comparable<CDependent<long long>>);
+static_assert(bf_three_way_comparable<CDependent<long long>>);
+#endif
+
+#if __cplusplus >= 201103L
+template<int W>
+struct D {
+  __int128 i : W;
```

Comment:
`__int128` is an extension, so I'd like us to avoid it in DR tests. Would `unsigned long long` suffice for this test?

Reply:
Bit-fields with width less than 32 are automatically promoted to `int` (https://stackoverflow.com/a/32529155), so if we tried `int64_t` and `int32_t` it would already work because of the promotion: https://godbolt.org/z/oMxE7oq31
(Also that might have been a clang bug because either that integral promotion shouldn't happen or it should have technically counted as narrowing before this DR because it is an implicit conversion that is narrowing via https://eel.is/c++draft/dcl.init.list#7.4.2)

So we need two types wider than `int`, and with a 32 bit int, that means a 64 bit and 128 bit type (Or a `_BitInt(33)` and `_BitInt(34)`, but that's also an extension in C++)

Reply:
Sorry, I still don't fully understand the situation, I guess.
> if we tried int64_t and int32_t it would already work because of the promotion

This holds for every standard integer types, because they are ranked. Extended integer types also have ranks. There's enough implementation-defined-ness in extended integer types to make me feel cautious seeing them in DR tests.

Reply:
Narrowing conversions don't consider the rank of integral types, only the range of values of the source and destination types. For `struct { long long i : 32; } x;` `int y{ x.i }`, clang uses an integral promotion (not an integral conversion) to initialize `y`, and it considers this to be non-narrowing even though it's from `long long` to `int`.

https://github.com/llvm/llvm-project/blob/ded04bf5d32a4fd5e0919053a598443f9d773549/clang/lib/Sema/SemaOverload.cpp#L549-L551

This leads to the situation where with `struct { __int128 i : 32; } x;`, `int{ x.i }` is not considered narrowing in clang, but `int64_t{ x.i }` *is*.

Reply:
Bear we me some more.
So you currently have the following tests that I presume test https://eel.is/c++draft/dcl.init.list#7.4.1 both positively and negatively:
```cpp
int64_t d1{ d<__int128, 63>().i }; // int64_t can represent all the values of a hypothetical 63-bit signed type, goo
int64_t d2{ d<__int128, 64>().i }; // same for 64 bits, good
int64_t d3{ d<__int128, 65>().i }; // 65 bits, 1 bit more that int64_t can represent
```
Why can't we "downgrade" this to
```cpp
int32_t d1{ d<int64_t, 31>().i };
int32_t d2{ d<int64_t, 32>().i };
int32_t d3{ d<int64_t, 33>().i };
```
Presumably this should yield the same result when https://eel.is/c++draft/dcl.init.list#7.4.1 is considered.

Reply:
Because that currently does a slightly different thing and currently does compile: https://godbolt.org/z/Ee61TPYnM (so the behaviour isn't affected by the defect report)

But it seems like that was because of a bug, so I think it's fine to get rid of __int128